// @remix-run/data-table/dist/lib/database.js
import { QueryBuilder } from "./database/query-builder.js";
import { DatabaseRuntime } from "./database/runtime.js";
/**
 * Creates a database runtime from an adapter.
 * @param adapter Adapter implementation responsible for SQL execution.
 * @param options Optional runtime options.
 * @param options.now Clock function used for auto-managed timestamps.
 * @returns A {@link Database} API instance.
 * @example
 * ```ts
 * import { column as c, createDatabase, table } from 'remix/data-table'
 *
 * let users = table({
 *   name: 'users',
 *   columns: {
 *     id: c.integer(),
 *     email: c.varchar(255),
 *   },
 * })
 *
 * let db = createDatabase(adapter)
 * let rows = await db.query(users).where({ id: 1 }).all()
 * ```
 */
export function createDatabase(adapter, options) {
    let now = options?.now ?? defaultNow;
    return new DatabaseRuntime({
        adapter,
        token: undefined,
        now,
        savepointCounter: { value: 0 },
    });
}
/**
 * Creates a database runtime bound to an existing adapter transaction token.
 * This is an internal helper used by the migration runner.
 * @param adapter Adapter implementation responsible for SQL execution.
 * @param token Active adapter transaction token.
 * @param options Optional runtime options.
 * @param options.now Clock function used for auto-managed timestamps.
 * @returns A {@link Database} API instance bound to the provided transaction.
 */
export function createDatabaseWithTransaction(adapter, token, options) {
    let now = options?.now ?? defaultNow;
    return new DatabaseRuntime({
        adapter,
        token,
        now,
        savepointCounter: { value: 0 },
    });
}
function defaultNow() {
    return new Date();
}
export { QueryBuilder };
