// pretty-dom@1.0.0/index.js
// index.js
import 'animate.css';

// 定义一个有趣的DOM操作方法
function addFunnyAnimation(selector, animationName) {
  const elements = document.querySelectorAll(selector);
  elements.forEach(el => {
    el.classList.add('animated', animationName);
  });
}

// 定义一个随机改变背景颜色的方法
function randomizeBackgroundColor(selector) {
  const elements = document.querySelectorAll(selector);
  elements.forEach(el => {
    el.style.backgroundColor = getRandomColor();
  });
}

// 生成随机颜色的函数
function getRandomColor() {
  const letters = '0123456789ABCDEF';
  let color = '#';
  for (let i = 0; i < 6; i++) {
    color += letters[Math.floor(Math.random() * 16)];
  }
  return color;
}

// 动态改变文字颜色的方法
function changeTextColor(selector, interval = 1000) {
  const colors = ['red', 'blue', 'green', 'purple', 'orange'];
  const elements = document.querySelectorAll(selector);
  elements.forEach(el => {
    setInterval(() => {
      el.style.color = colors[Math.floor(Math.random() * colors.length)];
    }, interval);
  });
}

// 实现有趣的鼠标跟随效果
function followCursor(selector, options = {}) {
  const defaultOptions = {
    followSpeed: 0.1, // 跟随速度
    maxDistance: 100 // 最大跟随距离
  };
  const settings = { ...defaultOptions, ...options };
  
  const elements = document.querySelectorAll(selector);
  elements.forEach(el => {
    const rect = el.getBoundingClientRect();
    let isFollowing = false;

    document.addEventListener('mousemove', (e) => {
      const x = e.clientX;
      const y = e.clientY;

      if (!isFollowing && (Math.abs(x - rect.left) < settings.maxDistance || Math.abs(y - rect.top) < settings.maxDistance)) {
        isFollowing = true;
        el.style.transition = 'transform 0.1s';
        el.style.transform = `translate(${(x - rect.left - window.scrollX) * settings.followSpeed}px, ${(y - rect.top - window.scrollY) * settings.followSpeed}px)`;
      }

      if (isFollowing && (Math.abs(x - rect.left) >= settings.maxDistance || Math.abs(y - rect.top) >= settings.maxDistance)) {
        isFollowing = false;
        el.style.transform = 'translate(0, 0)';
      }
    });
  });
}

// 导出模块的方法
export { addFunnyAnimation, randomizeBackgroundColor, changeTextColor, followCursor };
