// tsdown@0.1.1/dist/index.js
import {
  logger,
  removeFiles
} from "./chunk-J2CF23Y7.js";

// src/index.ts
import { rolldown } from "rolldown";

// src/options.ts
import { existsSync } from "node:fs";
import { stat } from "node:fs/promises";
import process from "node:process";
import path from "node:path";
import { globby } from "globby";
import { loadConfig } from "unconfig";

// src/error.ts
var PrettyError = class extends Error {
  constructor(message) {
    super(message);
    this.name = this.constructor.name;
    if (typeof Error.captureStackTrace === "function") {
      Error.captureStackTrace(this, this.constructor);
    } else {
      this.stack = new Error(message).stack;
    }
  }
};

// src/options.ts
async function normalizeOptions(options) {
  options = {
    ...await loadConfigFile(options),
    ...options
  };
  let {
    entry,
    format = ["es"],
    plugins = [],
    external = [],
    clean = false
  } = options;
  if (!entry || Object.keys(entry).length === 0) {
    throw new PrettyError(`No input files, try "tsdown <your-file>" instead`);
  }
  if (typeof entry === "string") {
    entry = [entry];
  }
  if (Array.isArray(entry)) {
    const resolvedEntry = await globby(entry);
    if (resolvedEntry.length > 0) {
      entry = resolvedEntry;
      logger.info(`Building entry: ${entry}`);
    } else {
      throw new PrettyError(`Cannot find ${entry}`);
    }
  } else {
    Object.keys(entry).forEach((alias) => {
      const filename = entry[alias];
      if (!existsSync(filename)) {
        throw new PrettyError(`Cannot find ${alias}: ${filename}`);
      }
    });
    logger.info(`Building entry: ${JSON.stringify(entry)}`);
  }
  if (!Array.isArray(format))
    format = [format];
  if (format.length === 0)
    format = ["es"];
  if (clean && !Array.isArray(clean))
    clean = [];
  return {
    entry,
    plugins,
    external,
    format,
    outDir: options.outDir || "dist",
    clean: clean ?? false
  };
}
async function loadConfigFile(options) {
  let { config: filePath } = options;
  if (filePath === false)
    return {};
  let cwd = process.cwd();
  let overrideConfig = false;
  let stats;
  if (typeof filePath === "string" && (stats = await stat(filePath).catch(() => null))) {
    const resolved = path.resolve(filePath);
    if (stats.isFile()) {
      overrideConfig = true;
      filePath = resolved;
      cwd = path.dirname(filePath);
    } else {
      cwd = resolved;
    }
  }
  const { config, sources } = await loadConfig({
    sources: overrideConfig ? [{ files: filePath, extensions: [] }] : [
      {
        files: "tsdown.config",
        extensions: ["ts", "mts", "cts", "js", "mjs", "cjs", "json", ""]
      },
      {
        files: "package.json",
        extensions: [],
        rewrite: (config2) => config2?.tsdown
      }
    ],
    cwd,
    defaults: {}
  });
  if (sources.length > 0) {
    logger.info(`Using tsdown config: ${sources.join(", ")}`);
  }
  return config;
}
function resolveFormat(format) {
  return format === "esm" ? "es" : format;
}

// src/index.ts
async function build(userOptions = {}) {
  const { entry, external, plugins, outDir, format, clean } = await normalizeOptions(userOptions);
  if (clean) {
    await removeFiles(["**/*", ...clean], outDir);
    logger.info("Cleaning output folder");
  }
  const build2 = await rolldown({
    input: entry,
    external,
    plugins
  });
  await Promise.all(
    format.map(
      (format2) => build2.write({
        format: resolveFormat(format2),
        dir: outDir
      })
    )
  );
  logger.info("Build complete");
}
function defineConfig(options) {
  return options;
}
export {
  build,
  defineConfig
};
