// @remix-run/data-table/dist/lib/database/runtime.js
import { DataTableAdapterError, DataTableQueryError } from "../errors.js";
import { isSqlStatement, rawSql } from "../sql.js";
import { getTableName } from "../table.js";
import { executeOperation } from "./execution-context.js";
import { asQueryTableInput, getPrimaryKeyWhere, getPrimaryKeyWhereFromRow, normalizeOrderByInput, resolveCreateRowWhere, toWriteResult, } from "./helpers.js";
import { createInitialQueryState, QueryBuilder } from "./query-builder.js";
export class DatabaseRuntime {
    #adapter;
    #token;
    #now;
    #savepointCounter;
    constructor(options) {
        this.#adapter = options.adapter;
        this.#token = options.token;
        this.#now = options.now;
        this.#savepointCounter = options.savepointCounter;
    }
    get adapter() {
        return this.#adapter;
    }
    now() {
        return this.#now();
    }
    query = (table) => new QueryBuilder(this, table, createInitialQueryState());
    async create(table, values, options) {
        let touch = options?.touch;
        let query = this.query(asQueryTableInput(table));
        if (options?.returnRow !== true) {
            let result = await query.insert(values, { touch });
            return toWriteResult(result);
        }
        if (this.#adapter.capabilities.returning) {
            let result = (await query.insert(values, {
                returning: '*',
                touch,
            }));
            let row = result.row;
            if (!row) {
                throw new DataTableQueryError('create({ returnRow: true }) failed to return an inserted row');
            }
            if (!options.with) {
                return row;
            }
            let where = getPrimaryKeyWhereFromRow(table, row);
            let loaded = await this.findOne(table, {
                where,
                with: options.with,
            });
            if (!loaded) {
                throw new DataTableQueryError('create({ returnRow: true }) failed to load inserted row');
            }
            return loaded;
        }
        let insertResult = await query.insert(values, { touch });
        let where = resolveCreateRowWhere(table, values, toWriteResult(insertResult).insertId);
        let loaded = await this.findOne(table, {
            where,
            with: options.with,
        });
        if (!loaded) {
            throw new DataTableQueryError('create({ returnRow: true }) failed to load inserted row');
        }
        return loaded;
    }
    async createMany(table, values, options) {
        let query = this.query(asQueryTableInput(table));
        if (options?.returnRows === true) {
            if (!this.#adapter.capabilities.returning) {
                throw new DataTableQueryError('createMany({ returnRows: true }) is not supported by this adapter');
            }
            let result = (await query.insertMany(values, {
                returning: '*',
                touch: options.touch,
            }));
            return result.rows;
        }
        let result = await query.insertMany(values, {
            touch: options?.touch,
        });
        return toWriteResult(result);
    }
    async find(table, value, options) {
        if (value == null) {
            return null;
        }
        let query = this.query(asQueryTableInput(table));
        if (options?.with) {
            return query.with(options.with).find(value);
        }
        return query.find(value);
    }
    async findOne(table, options) {
        let query = this.query(asQueryTableInput(table)).where(options.where);
        let orderBy = normalizeOrderByInput(options.orderBy);
        for (let [column, direction] of orderBy) {
            query = query.orderBy(column, direction);
        }
        if (options.with) {
            return query.with(options.with).first();
        }
        return query.first();
    }
    async findMany(table, options) {
        let query = this.query(asQueryTableInput(table));
        if (options?.where) {
            query = query.where(options.where);
        }
        let orderBy = normalizeOrderByInput(options?.orderBy);
        for (let [column, direction] of orderBy) {
            query = query.orderBy(column, direction);
        }
        if (options?.limit !== undefined) {
            query = query.limit(options.limit);
        }
        if (options?.offset !== undefined) {
            query = query.offset(options.offset);
        }
        if (options?.with) {
            return query.with(options.with).all();
        }
        return query.all();
    }
    async count(table, options) {
        let query = this.query(asQueryTableInput(table));
        if (options?.where) {
            query = query.where(options.where);
        }
        return query.count();
    }
    async update(table, value, changes, options) {
        let where = getPrimaryKeyWhere(table, value);
        if (this.#adapter.capabilities.returning) {
            let updateResult = (await this.query(asQueryTableInput(table)).where(where).update(changes, {
                touch: options?.touch,
                returning: '*',
            }));
            let updatedRow = updateResult.rows[0];
            if (!updatedRow) {
                throw new DataTableQueryError('update() failed to find row for table "' + getTableName(table) + '"');
            }
            if (!options?.with) {
                return updatedRow;
            }
            let loaded = await this.findOne(table, {
                where: getPrimaryKeyWhereFromRow(table, updatedRow),
                with: options.with,
            });
            if (!loaded) {
                throw new DataTableQueryError('update() failed to find row for table "' + getTableName(table) + '"');
            }
            return loaded;
        }
        await this.query(asQueryTableInput(table)).where(where).update(changes, {
            touch: options?.touch,
        });
        let loaded = await this.find(table, value, { with: options?.with });
        if (!loaded) {
            throw new DataTableQueryError('update() failed to find row for table "' + getTableName(table) + '"');
        }
        return loaded;
    }
    async updateMany(table, changes, options) {
        let query = this.query(asQueryTableInput(table)).where(options.where);
        let orderBy = normalizeOrderByInput(options.orderBy);
        for (let [column, direction] of orderBy) {
            query = query.orderBy(column, direction);
        }
        if (options.limit !== undefined) {
            query = query.limit(options.limit);
        }
        if (options.offset !== undefined) {
            query = query.offset(options.offset);
        }
        let result = await query.update(changes, { touch: options.touch });
        return toWriteResult(result);
    }
    async delete(table, value) {
        let where = getPrimaryKeyWhere(table, value);
        let result = await this.query(asQueryTableInput(table)).where(where).delete();
        return toWriteResult(result).affectedRows > 0;
    }
    async deleteMany(table, options) {
        let query = this.query(asQueryTableInput(table)).where(options.where);
        let orderBy = normalizeOrderByInput(options.orderBy);
        for (let [column, direction] of orderBy) {
            query = query.orderBy(column, direction);
        }
        if (options.limit !== undefined) {
            query = query.limit(options.limit);
        }
        if (options.offset !== undefined) {
            query = query.offset(options.offset);
        }
        let result = await query.delete();
        return toWriteResult(result);
    }
    async exec(statement, values = []) {
        let sqlStatement = isSqlStatement(statement) ? statement : rawSql(statement, values);
        return this[executeOperation]({
            kind: 'raw',
            sql: sqlStatement,
        });
    }
    async transaction(callback, options) {
        if (!this.#token) {
            let token = await this.#adapter.beginTransaction(options);
            let tx = new DatabaseRuntime({
                adapter: this.#adapter,
                token,
                now: this.#now,
                savepointCounter: this.#savepointCounter,
            });
            try {
                let result = await callback(tx);
                await this.#adapter.commitTransaction(token);
                return result;
            }
            catch (error) {
                await this.#adapter.rollbackTransaction(token);
                throw error;
            }
        }
        if (!this.#adapter.capabilities.savepoints) {
            throw new DataTableQueryError('Nested transactions require adapter savepoint support');
        }
        let savepointName = 'sp_' + String(this.#savepointCounter.value);
        this.#savepointCounter.value += 1;
        await this.#adapter.createSavepoint(this.#token, savepointName);
        try {
            let result = await callback(this);
            await this.#adapter.releaseSavepoint(this.#token, savepointName);
            return result;
        }
        catch (error) {
            await this.#adapter.rollbackToSavepoint(this.#token, savepointName);
            await this.#adapter.releaseSavepoint(this.#token, savepointName);
            throw error;
        }
    }
    async [executeOperation](operation) {
        try {
            return await this.#adapter.execute({
                operation,
                transaction: this.#token,
            });
        }
        catch (error) {
            throw new DataTableAdapterError('Adapter execution failed', {
                cause: error,
                metadata: {
                    dialect: this.#adapter.dialect,
                    operationKind: operation.kind,
                },
            });
        }
    }
}
